tests/mr/device/arena_mr_tests.cpp

```diff
@@ -0,0 +1,519 @@
+{
+  ga_->allocate(superblock::minimum_size + 256);
+  ga_->allocate(superblock::minimum_size + 256);
+  EXPECT_EQ(ga_->allocate(superblock::minimum_size + 256), nullptr);
+}
+
+TEST_F(ArenaTest, GlobalArenaDeallocate)  // NOLINT
+{
+  auto* ptr = ga_->allocate(superblock::minimum_size * 2);
+  EXPECT_EQ(ptr, fake_address3);
+  ga_->deallocate(ptr, superblock::minimum_size * 2, {});
+  ptr = ga_->allocate(superblock::minimum_size * 2);
+  EXPECT_EQ(ptr, fake_address3);
+}
+
+TEST_F(ArenaTest, GlobalArenaDeallocateAlignUp)  // NOLINT
+{
+  auto* ptr  = ga_->allocate(superblock::minimum_size + 256);
+  auto* ptr2 = ga_->allocate(superblock::minimum_size + 512);
+  ga_->deallocate(ptr, superblock::minimum_size + 256, {});
+  ga_->deallocate(ptr2, superblock::minimum_size + 512, {});
+  EXPECT_EQ(ga_->allocate(arena_size_), fake_address3);
+}
+
+TEST_F(ArenaTest, GlobalArenaDeallocateFromOtherArena)  // NOLINT
+{
+  auto sb      = ga_->acquire(512);
+  auto const b = sb.first_fit(512);
+  ga_->release(std::move(sb));
+  ga_->deallocate_from_other_arena(b.pointer(), b.size());
+  EXPECT_EQ(ga_->allocate(arena_size_), fake_address3);
+}
+
+/**
+ * Test arena.
+ */
+
+TEST_F(ArenaTest, ArenaAllocate)  // NOLINT
+{
+  EXPECT_EQ(a_->allocate(superblock::minimum_size), fake_address3);
+  EXPECT_EQ(a_->allocate(256), fake_address4);
+}
+
+TEST_F(ArenaTest, ArenaDeallocate)  // NOLINT
+{
+  auto* ptr = a_->allocate(superblock::minimum_size);
+  a_->deallocate(ptr, superblock::minimum_size, {});
+  auto* ptr2 = a_->allocate(256);
+  a_->deallocate(ptr2, 256, {});
+  EXPECT_EQ(a_->allocate(superblock::minimum_size), fake_address3);
+}
+
+TEST_F(ArenaTest, ArenaDeallocateMergePrevious)  // NOLINT
+{
+  auto* ptr  = a_->allocate(256);
+  auto* ptr2 = a_->allocate(256);
+  a_->allocate(256);
+  a_->deallocate(ptr, 256, {});
+  a_->deallocate(ptr2, 256, {});
+  EXPECT_EQ(a_->allocate(512), fake_address3);
+}
+
+TEST_F(ArenaTest, ArenaDeallocateMergeNext)  // NOLINT
+{
+  auto* ptr  = a_->allocate(256);
+  auto* ptr2 = a_->allocate(256);
+  a_->allocate(256);
+  a_->deallocate(ptr2, 256, {});
+  a_->deallocate(ptr, 256, {});
+  EXPECT_EQ(a_->allocate(512), fake_address3);
+}
+
+TEST_F(ArenaTest, ArenaDeallocateMergePreviousAndNext)  // NOLINT
+{
+  auto* ptr  = a_->allocate(256);
+  auto* ptr2 = a_->allocate(256);
+  a_->deallocate(ptr, 256, {});
+  a_->deallocate(ptr2, 256, {});
+  EXPECT_EQ(a_->allocate(2_KiB), fake_address3);
+}
+
+TEST_F(ArenaTest, ArenaDefragment)  // NOLINT
+{
+  std::vector<void*> pointers;
+  std::size_t num_pointers{4};
+  for (std::size_t i = 0; i < num_pointers; i++) {
+    pointers.push_back(a_->allocate(superblock::minimum_size));
+  }
+  for (auto* ptr : pointers) {
+    a_->deallocate(ptr, superblock::minimum_size, {});
+  }
+  EXPECT_EQ(ga_->allocate(arena_size_), nullptr);
+  a_->defragment();
+  EXPECT_EQ(ga_->allocate(arena_size_), fake_address3);
+}
+
+/**
+ * Test arena_memory_resource.
+ */
```

Comment:
I would argue this is the level we should be testing. Behaviors specific to the MR type should be tested here. General usability of the MR through the `device_memory_resource` interface is tested in `mr_test.hpp/cpp`.